nailgun-server/src/main/java/com/martiansoftware/nailgun/NGInputStream.java

```diff
@@ -40,15 +42,16 @@
     private final DataInputStream din;
     private InputStream stdin = null;
     private boolean eof = false;
-    private long remaining = 0;
+    private boolean clientConnected = true;
+    private int remaining = 0;
     private byte[] oneByteBuffer = null;
     private final DataOutputStream out;
     private boolean started = false;
-    private long lastReadTime = System.currentTimeMillis();
-    private final Future readFuture;
-    private final Set clientListeners = new HashSet();
-    private final Set heartbeatListeners = new HashSet();
+    private AtomicLong lastReadTime = new AtomicLong(System.currentTimeMillis());
```

Comment:
https://javaantipatterns.wordpress.com/2012/03/15/measure-time-intervals-using-system-currenttimemillis/

Reply:
As it existed before, will change in some next diff

Reply:
Actually, let me change it. It is only used to report back heartbeats and this is incorrect anyways because it gives back time between any data communications, not just heartbeats